sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/GcsUtilIT.java

```diff
@@ -76,4 +86,67 @@ public void testRewriteMultiPart() throws IOException {
 
     gcsUtil.remove(Lists.newArrayList(dstFilename));
   }
+
+  @Test
+  public void testWriteAndReadGcsWithGrpc() throws IOException {
+    TestPipelineOptions options =
+        TestPipeline.testingPipelineOptions().as(TestPipelineOptions.class);
+
+    // set the experimental flag to enable grpc
+    ExperimentalOptions experimental = options.as(ExperimentalOptions.class);
+    experimental.setExperiments(Collections.singletonList("use_grpc_for_gcs"));
+
+    GcsOptions gcsOptions = options.as(GcsOptions.class);
+    GcsUtil gcsUtil = gcsOptions.getGcsUtil();
+    assertNotNull(gcsUtil);
+
+    String testContent = "This is a test string.";
+
+    // Write a test file in a bucket without gRPC enabled.
+    // This assumes that GCS gRPC feature is not enabled in every bucket by default.
+    // If the following assertion fails, we can revisit the GA status of this feature and check
+    // whether we can remove the assertion.
+    String tempLocationWithoutGrpc = "gs://temp-storage-for-end-to-end-tests-cmek/temp";
+    String wrongFilename =
+        String.format(
+            "%s/GcsUtilIT-%tF-%<tH-%<tM-%<tS-%<tL.testWriteAndReadGcsWithGrpc.txt",
+            tempLocationWithoutGrpc, new Date());
+    assertThrows(IOException.class, () -> writeGcsTextFile(gcsUtil, wrongFilename, testContent));
+
+    // Write a test file in a bucket with gRPC enabled.
+    String tempLocationWithGrpc = "gs://gcs-grpc-team-apache-beam-testing/temp";
```

Comment:
Let's make this configurable. Define this as a static final field, and detect whether a system property (e.g. `gcs_temp_bucket_grpc` exists) which can override the default.

Reply:
Done.